gpflow/features.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+import tensorflow as tf
+from functools import singledispatch
+from abc import ABCMeta, abstractmethod
+
+from .param import Param, Parameterized
+from . import conditionals
+from ._settings import settings
+
+
+float_type = settings.dtypes.float_type
+
+
+class InducingFeature(Parameterized, metaclass=ABCMeta):
+    """
+    Abstract base class for inducing features.
+    """
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Returns the number of features, relevant for example to determine the
+        size of the variational distribution.
+        """
+        pass
+
+    @abstractmethod
+    def Kuu(self, kern):
+        """
+        Calculates the covariance matrix between features for kernel `kern`.
+        """
+        pass
+
+    @abstractmethod
+    def Kuf(self, kern, Xnew):
+        """
+        Calculates the covariance matrix with function values at new points
+        `Xnew` for kernel `kern`.
+        """
+        pass
+
+
+class InducingPoints(InducingFeature):
+    """
+    Real-space inducing points
+    """
+
+    def __init__(self, Z):
+        """
+        :param Z: the initial positions of the inducing points, size M x D
+        """
+        super().__init__()
+        self.Z = Param(Z)
+        self._Mfeat = len(Z)
+
+    def __len__(self):
+        return self._Mfeat
+
+    def Kuu(self, kern):
+        Kzz = kern.K(self.Z)
+        Kzz += settings.numerics.jitter_level * tf.eye(self._Mfeat, dtype=float_type)
+        return Kzz
+
+    def Kuf(self, kern, Xnew):
+        Kzx = kern.K(self.Z, Xnew)
+        return Kzx
+
+
+@singledispatch
+def conditional(feat, kern, Xnew, f, full_cov=False, q_sqrt=None, whiten=False):
+    """
+    Note the changed function signature compared to conditionals.conditional() to allow for dispatch.
+    """
+    raise NotImplementedError("No implementation for {} found".format(type(feat).__name__))
+
+@conditional.register(InducingPoints)
```

Comment:
Does GPflow need the singledispatch construction here? Can't all `InducingFeature` objects not use the same `feature_conditional` function?

I do remember discussing some things about possibly supporting matrices that have more structure, which makes them easier to invert. Is that the rationale behind this decision?

Reply:
Yes, depending on the feature, the structure of the code in `conditional()` may be more efficient one way or another. If you only have a single `feature_conditional` function, it won't be possible to do that, and I'd have to end up maintaining my own branch of code. This way, with a new feature you can simply register another function.

We could change the function name in the following row to something else, then you can just re-register the same function for different types. Or, if they're in here already, just add a `@conditional.register(Multiscale)` here.

Reply:
Great, that confirms why it's implemented like that.

Why does the function name need to be changed to something other than `_`? Can we not just do the following?
```
@conditional.register(InducingPoints)
@conditional.register(Multiscale)
def _(feat, kern, Xnew, f, full_cov=False, q_sqrt=None, whiten=False):
```

Reply:
That's what I meant. You only need to change it if you want to be able to register from outside this module, e.g. if some third-party module wants to do `@gpflow.features.conditional.register(MyOwnFeature)(gpflow.features.plain_features_conditional)` ...

Reply:
Ah yes! Looks like a great change.